### PR TITLE
Fix report blocking to use the last five resolved reports

### PR DESCRIPTION
**Pull request by Blackspirits**
## Summary

Fix temporary report blocking so it evaluates the last five resolved reports, matching the original intended behavior.

The current implementation counts dismissed reports within the last week and only blocks when that count is exactly five. This can incorrectly stop blocking after a sixth dismissed report, and can also block when five dismissed reports are followed by a more recent upheld report.

Apply the same last-five logic to both reporter-level blocking and item-level blocking. The five most recent resolved reports must all be dismissed and all be within the last week.

## Validation

* Added regression coverage confirming that more than five consecutive dismissed reports remain blocked.
* Added regression coverage confirming that a more recent upheld report prevents blocking.
* Added regression coverage confirming that the last five resolved reports must all be within the one-week window.
* Covered both reporter-level and item-level report blocking.

### CODE REVIEW

File: app/controllers/reports_controller.rb [hunk 1]
```diff
@@ -273,8 +273,11 @@ def check_for_blocked_report(report = nil)
   end
 
   def item_reporting_blocked_until
-    recent_reports = Report.resolved.where(item:, created_at: 1.week.ago..).order(:created_at)
-    recent_reports.first.created_at + 1.week if recent_reports.count(&:dismissed?) == 5
+    recent_reports = Report.resolved.where(item:).order(created_at: :desc, id: :desc).limit(5).to_a
```
Comment (JasonBarnabe) [inline app/controllers/reports_controller.rb]: Since we only want to consider the case where the oldest of the 5 is within the past week, why are you removing the ActiveRecord condition that we load only reports created in the last week? Generally it's better to check these things in the DB than load the records and check them in Ruby.

Comment (JasonBarnabe) [inline app/controllers/reports_controller.rb]: Why switch from sorting by `:created_at` and getting the first record to sorting by `created_at :desc` and getting the last?

File: test/system/reports/blocked_test.rb [hunk 2]
```diff
@@ -12,6 +12,35 @@ class ReportsBlockedTest < ApplicationSystemTestCase
     assert_content 'Due to recent reports'
   end
 
+  test 'reporting remains blocked after more than five dismissed reports' do
```
Comment (JasonBarnabe) [inline test/system/reports/blocked_test.rb]: I don't it's good to have system tests for details like this. System tests are slow and should generally just do the most common cases.

If you're looking for somewhere to test the `ReportsController` code then that may mean that it should be refactored into the `Report` model so it can be tested more easily.

**Review by JasonBarnabe — CHANGES_REQUESTED** [review]

File: app/controllers/reports_controller.rb (hunk 1 above)
Comment (Blackspirits) [reply app/controllers/reports_controller.rb]: Good point. I restored the one-week condition in the ActiveRecord query, so reports older than a week are filtered by the database before loading the five candidates. I also moved the shared logic into Report.reporting_blocked_until.

Comment (Blackspirits) [reply app/controllers/reports_controller.rb]: The descending order is intentional once LIMIT 5 is applied: it selects the five most recent resolved reports. last is then the oldest report within those five, which determines when the block expires.

File: test/system/reports/blocked_test.rb (hunk 2 above)
Comment (Blackspirits) [reply test/system/reports/blocked_test.rb]: Agreed. I removed the detailed system tests and moved the shared blocking logic into the Report model, with the edge cases covered in ReportTest. The existing common-case system test remains.

File: app/models/report.rb [hunk 3]
```diff
@@ -82,6 +82,13 @@ class Report < ApplicationRecord
   validates :discussion_category, presence: true, if: -> { reason == REASON_WRONG_CATEGORY }
   validates :private_explanation, length: { maximum: 65_535 }
 
+  def self.reporting_blocked_until(scope)
```
Comment (JasonBarnabe) [inline app/models/report.rb]: Because you're always passing in a `Report` AR collection and this method is a class method on `Report`, you can instead drop the argument and use `self`, then the callers can use e.g. `Report.where(item:).reporting_blocked_until`.

File: app/controllers/reports_controller.rb [hunk 4]
```diff
@@ -273,8 +273,7 @@ def check_for_blocked_report(report = nil)
   end
 
   def item_reporting_blocked_until
```
Comment (JasonBarnabe) [inline app/controllers/reports_controller.rb]: Because this method is now a one-liner and is only called from one place, we can drop it and just have the caller call `Report.reporting_blocked_until` directly.

**Review by JasonBarnabe — CHANGES_REQUESTED** [review]

File: app/models/report.rb (hunk 3 above)
Comment (Blackspirits) [reply app/models/report.rb]: Updated. reporting_blocked_until no longer takes a relation argument and is now called on scoped Report relations; the model tests exercise the relation API directly as well.

File: app/controllers/reports_controller.rb (hunk 4 above)
Comment (Blackspirits) [reply app/controllers/reports_controller.rb]: Updated. I removed the one-line item_reporting_blocked_until wrapper and call Report.where(item:).reporting_blocked_until directly from check_for_blocked_report.
